File: rust_src/src/lib/problem007.rs
/* Project Euler Problem 7 Solution
 *
 * Problem statement:
 * By listing the first six prime numbers: 2, 3, 5, 7, 11,
 * and 13, we can see that the 6th prime is 13. What is the
 * 10 001st prime number?
 *
 * Solution description:
 * This script uses a brute-force approach to determine the
 * n'th prime number. This is a slow approach which should
 * be improved in the future!
 *
 * Author: Daniel Schuette
 * Date: 2018/02/27
 * License: MIT (see ../LICENSE.md)
 */
use crate::utils;

// Returns the n'th prime number.
pub fn nth_prime(n: i64) -> i64 {
    let mut prime_count: i64 = 1; /* 2 is already counted in */
    let mut current_num: i64 = 3;

    loop {
        if utils::is_prime(current_num) {
            prime_count += 1;
            if prime_count == n {
                return current_num;
            }
        }
        current_num += 2;
    }
}

File: cpp_src/problem007.cpp
// Project Euler Problem 007 Solution
//
// Problem statement: By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.What is the 10 001st prime number?
//
// Solution description:
//
// Author: Tom Praschan
// 2019/03/03
// License: MIT (see ../LICENSE.md)

#include <omp.h>
#include <fmt/format.h>
#include "prime_utils.hpp"

//Computes the nth prime number
size_t nth_prime(size_t n) {
    if (n == 1) {
        return 2;
    } else if (n == 2) {
        return 3;
    }
    size_t index = 2, p = 3;
    while (index != n) {
        p += 2;
        if (is_prime(p)) {
            index++;
        }
    }
    return p;
}

#ifndef TESTING
int main(int argc, char **argv) {
   auto start = omp_get_wtime();
   auto solution = nth_prime(10001);
   auto end = omp_get_wtime();

   fmt::print("Solution: {} \n", solution);
   fmt::print("Elapsed time: {}s \n", end - start);
}
#endif
